tcl/return.tcl

```diff
@@ -3,54 +3,68 @@ namespace eval qc {
 }
 
 proc qc::return2client { args } {
-    #| Return data to http client
-    # Usage return2client ?code code? ?content-type mime-type? ?file file? ?filename filename? ?html html? ?text text? ?xml xml? ?json json? ?filter_cc boolean? ?header header? .. 
+    #| Return data to http client if a connection exists otherwise just output the given content.
+    # Usage return2client ?code code? ?content-type mime-type? ?html html? ?text text? ?xml xml? ?json json? ?file file? ?filename filename? ?download boolean? ?filter_cc boolean? ?header header? .. 
     set arg_names [qc::args2vars $args]
-    set headers [lexclude $arg_names file filename html xml text json code content-type]
+    set headers [lexclude $arg_names file filename html xml text json file filename download code content-type filter_cc]
     default code 200
     default filter_cc no
+    default filename [qc::url_path [ns_conn url]]
 
     if { [info exists file] } {
-        default filename [qc::url_path [ns_conn url]]
+        # File
+        default download yes
         default content-type [qc::mime_type_guess $filename]
-        default content-disposition "attachment; $filename"
-        set headers [qc::lunion $headers [list "content-disposition"]]
-        set f [open $file rb]
-        set content [read $f]
-        close $f
-        set var content
-
+        set var file
     } elseif { [info exists html] } {
+        # HTML payload
+        default download no
 	default content-type "text/html; charset=utf-8"
 	set var html
-
     } elseif { [info exists xml] } {
+        # XML payload
+        default download no
 	default content-type "text/xml; charset=utf-8"
 	set var xml
-
     } elseif { [info exists text] } {
+        # Text payload
+        default download no
 	default content-type "text/plain; charset=utf-8"
 	set var text
-
     } elseif { [info exists json] } {
+        # JSON payload
+        default download no
 	default content-type "application/json; charset=utf-8"
 	set var json
-
     } else {
-	error "No payload given in html or xml or text or json" 
+        # Unknown payload
+	error "No payload given in html or xml or text or json or file" 
     }
 
-    if { $filter_cc } {
-	# mask credit card numbers
-	set $var [qc::format_cc_masked_string [set $var]] 
-    }
-    # Other headers
-    foreach name $headers {
-    	set conn_headers [ns_conn outputheaders]
-	ns_set update $conn_headers $name [set $name]
+    # Update headers and return payload to client
+    if { [expr 0x1 & [ns_conn flags]] } {
+        # no open connection
+        return [lindex $args end]
+    } else {
+        if { $filter_cc } {
+            # mask credit card numbers
+            set $var [qc::format_cc_masked_string [set $var]] 
+        }
```

Comment:
If `filter_cc` is specified and no connection exists, shouldn't we filter the return data?